packages/neon-api/src/transaction/validator.ts

```diff
@@ -0,0 +1,229 @@
+import { tx, rpc, u } from "@cityofzion/neon-core";
+import { getNetworkFee, getScriptHashesFromTxWitnesses } from "./util";
+
+export enum ValidationAttributes {
+  None = 0,
+  ValidUntilBlock = 1 << 0,
+  SystemFee = 1 << 1,
+  NetworkFee = 1 << 2,
+  Scripts = 1 << 3
+}
+
+export interface ValidationSuggestion<T> {
+  /**
+   * Whether this is auto-fixed by validator
+   * Validator will try to fix the transaction when param `autoFix` is set to TRUE
+   */
+  fixed: boolean;
+  prev: T;
+  suggestion: T;
+}
+
+export interface ValidationResult {
+  /**
+   * Whether the transaction is valid after validation
+   */
+  valid: boolean;
+  validaUntilBlock?: ValidationSuggestion<number>;
+  systemFee?: ValidationSuggestion<u.Fixed8>;
+  networkFee?: ValidationSuggestion<u.Fixed8>;
+  witnesses?: {
+    notSignedHash: string;
+  };
+}
+
+/**
+ * A class with functions to validate transaction
+ */
+export class TransactionValidator {
+  /**
+   * Transaction will be validated on this rpc node
+   */
+  public rpcClient: rpc.RPCClient;
+
+  /**
+   * Transaction that will be validated
+   */
+  public transaction: tx.Transaction;
+
+  constructor(rpc: rpc.RPCClient, transaction: tx.Transaction) {
+    this.rpcClient = rpc;
+    this.transaction = transaction;
+  }
+
+  /**
+   * validate transaction attribute - validUntilBlock
+   * @param autoFix will automatically fix transaction if specified as true
+   */
+  validateValidUntilBlock = async (
+    autoFix = false
+  ): Promise<ValidationResult> => {
+    const { validUntilBlock } = this.transaction;
+    const height = await this.rpcClient.getBlockCount();
+    if (
+      validUntilBlock <= height ||
+      validUntilBlock >= height + tx.Transaction.MAX_TRANSACTION_LIFESPAN
+    ) {
+      const vub_suggestion =
+        tx.Transaction.MAX_TRANSACTION_LIFESPAN + height - 1;
+      if (autoFix) {
+        this.transaction.validUntilBlock = vub_suggestion;
+        return {
+          valid: true,
+          validaUntilBlock: {
+            fixed: true,
+            prev: validUntilBlock,
+            suggestion: vub_suggestion
+          }
+        };
+      }
+      return {
+        valid: false,
+        validaUntilBlock: {
+          fixed: false,
+          prev: validUntilBlock,
+          suggestion: vub_suggestion
+        }
+      };
+    }
+    return {
+      valid: true
+    };
+  };
+
+  // validateIntents = async (): Promise<ValidationResult> => {
+  //   return {
+  //     result: true,
+  //     fixed: false
+  //   };
+  // };
+
+  /**
+   * validate systemFee
+   * @param autoFix will automatically fix transaction if specified as true
+   */
+  validateSystemFee = async (autoFix = false): Promise<ValidationResult> => {
```

Comment:
1. You might want to do a check on the existing systemFee to make sure that it is in an acceptable format too (integer without decimals).

2. Consider including the action to verify the intent here too. If the intent is invalid, the systemFee validation will also probably be worthless.

Reply:
1. Yeah, thanks for reminding.
2. OK, will do. Planning to using manifest

Reply:
Im not clear if in neo3 the VM will error out on scripts requiring verification. If it does, then it might be hard to validate the script and estimate the fee.

Reply:
Not very clear about your concern, I have added `validateIntents` using ContractManifest.